### PR TITLE
ISSUE-33802: ensure that deployment config will reload its data, if t…

### LINKED ISSUE #33802: Integration tests failing during Magento setup
### Preconditions (*)
1. Fresh Magento 2.4.3 installation
2. PHP 7.3 or PHP 7.4
3. Composer version 2.1.5 (also tried with the highest composer 1 version
4. no RabbitMQ

### Steps to reproduce (*)
1. Basically I follow the steps in this doc https://devdocs.magento.com/guides/v2.4/test/integration/integration_test_execution.html
2. `composer create-project --repository-url=https://repo.magento.com/ magento/project-community-edition`
3. `bin/magento module:enable --all`
4. `update the configuration in `dev/tests/integration/etc/install-config-mysql.php` (no RabbitMQ configuration)
5. `mysql -e "DROP DATABASE IF EXISTS magento_integration_tests"`
6. `mysql -e "CREATE DATABASE magento_integration_tests"`
7. `cd dev/tests/integration && php ../../../vendor/bin/phpunit`

### Expected result (*)
1. Magento test instance should be configured in dev/tests/integration/tmp/sandbox-{hash} dir
2. Tests defined in `dev/tests/integration/phpunit.xml.dist` should start running

### Actual result (*)
1. When running magento installation command, this command fails during the step "Disabling Maintenance Mode" with an error
```
In ResourceConnection.php line 148:

  [DomainException]
  Connection "default" is not defined
```
Log file is attached below

---
Please provide [Severity](https://devdocs.magento.com/guides/v2.4/contributor-guide/contributing.html#backlog) assessment for the Issue as Reporter. This information will help during Confirmation and Issue triage processes.

- [ ] Severity: **S0** _- Affects critical data or functionality and leaves users without workaround._
- [ ] Severity: **S1** _- Affects critical data or functionality and forces users to employ a workaround._
- [x] Severity: **S2** _- Affects non-critical data or functionality and forces users to employ a workaround._
- [ ] Severity: **S3** _- Affects non-critical data or functionality and does not force users to employ a workaround._
- [ ] Severity: **S4** _- Affects aesthetics, professional look and feel, “quality” or “usability”._
[log.txt](https://github.com/magento/magento2/files/6988301/log.txt)

### PR DESCRIPTION
**Pull request by sergiynezbritskiy**
ensure that deployment config will reload its data, if the key was not found

### Description (*)
Reset DeploymentConfig data if the requested configuration doesn't exist in cached data 

### Fixed Issues (if relevant)
1. Fixes magento/magento2#33802

### Manual testing scenarios (*)
1. Described in magento/magento2#33802

### Contribution checklist (*)
 - [x] Pull request has a meaningful description of its purpose
 - [x] All commits are accompanied by meaningful commit messages
 - [x] All new or changed code is covered with unit/integration tests (if applicable)
 - [x] README.md files for modified modules are updated and included in the pull request if any [README.md predefined sections](https://github.com/magento/devdocs/wiki/Magento-module-README.md) require an update
 - [x] All automated tests passed successfully (all builds are green)

### CODE REVIEW

**Comment by sergeynezbritskiy** [comment]
@magento run all tests

**Comment by sergeynezbritskiy** [comment]
@magento run all tests

**Comment by sergeynezbritskiy** [comment]
@magento run all tests

**Comment by sergeynezbritskiy** [comment]
@magento run Functional Tests B2B

**Comment by sergeynezbritskiy** [comment]
@magento run Functional Tests B2B

**Comment by sergeynezbritskiy** [comment]
@magento run Functional Tests B2B

**Comment by sergeynezbritskiy** [comment]
@magento run Functional Tests B2B

**Comment by sergeynezbritskiy** [comment]
Every run fails due to different failures, what's going on here ?

**Comment by BorisovskiP** [comment]
Hey @sergeynezbritskiy, we added your changes as a patch and in our case this indeed fixes the issue.

**Comment by oneserv-heuser** [comment]
@sergeynezbritskiy @BorisovskiP we did the same and can also confirm this

**Comment by tniebergall** [comment]
@sergeynezbritskiy thank you for this fix, we had the same issue during an upgrade vom 2.4.2-p1 to 2.4.3

**Review by ihor-sviziev — APPROVED** [review]

**Comment by magento-engcom-team** [comment]
Hi @ihor-sviziev, thank you for the review.
[ENGCOM-9239](https://jira.corp.magento.com/browse/ENGCOM-9239) has been created to process this Pull Request

**Review by sidolov — APPROVED** [review]

**Comment by Sanjay-100** [comment]
> heavy_check_mark QA Passed
> 
> **Manual testing scenario:**
> 
> 1. Installed a fresh Magento2.4-develop instance, and make sure that is working fine
> 2. Delete the database completely and re-create the same database name (That will be empty database as database deleted & re created)
> 3. Remove the file /app/etc/env.php
> 4. Re-install the with the database creation command i.e. `php bin/magento setup:install --backend-frontname="admin" --session-save="files" --db-host="localhost" --db-name="magento24develop" --db-user="root" --db-password="" --base-url="http://local.magento24develop.com/" --base-url-secure="https://local.magento24develop.com/" --admin-user="admin" --admin-password="admin123" --admin-email="admin@admin.com" --admin-firstname="Admin" --admin-lastname="Admin"`
> 
> **Before:** heavy_multiplication_x Getting an error Connection default not defined while executing the installation command
> 
> image
> 
> **After:** heavy_check_mark Successfully installing using the command
> 
> image
> 
> Worked on the install with the PR changes updates. Magento features are working fine. Since the issue is relevant to the installation setup and has no impact on any other features of it, so no additional regression is required.

I have already one magento setup now I just change the new database 
steps I did
- fresh copy Magento 2.3
- blank dtabase
- deleted env.php
- then run setup the command 


set Disabling Maintenance Mode:

In ResourceConnection.php line 148:
                                       
  Connection "default" is not defined

**Comment by ihor-sviziev** [comment]
Hi @sergeynezbritskiy,
Could you please resolve the conflict?

**Comment by cmuench** [comment]
I created a patch file for Magento 2.4.4 with the changes of this PR.
No warranty.
The patch fixed the issue in my CI pipelines.

https://gist.github.com/cmuench/de27258f0b4ccb8df3dee2e8507e935d

**Comment by norgeindian** [comment]
Has anybody by chance already a patch file for 2.4.5?

**Comment by ihor-sviziev** [comment]
@magento run Static Tests, Unit Tests

File: lib/internal/Magento/Framework/App/DeploymentConfig.php [hunk 1]
```diff
@@ -236,4 +240,26 @@ private function flattenParams(array $params, $path = null, array &$flattenResul
 
         return $flattenResult;
     }
+
+    /**
```
Comment (ihor-sviziev) [inline lib/internal/Magento/Framework/App/DeploymentConfig.php:244]:
```suggestion
    /**
     * Returns flat data by key
     *
```

File: lib/internal/Magento/Framework/App/DeploymentConfig.php [hunk 2]
```diff
@@ -236,4 +240,26 @@ private function flattenParams(array $params, $path = null, array &$flattenResul
...
+    private function getByKey(?string $key)
+    {
+        if (array_key_exists($key, $this->flatData) && $this->flatData[$key] === null) {
+            return '';
+        }
+
+        return $this->flatData[$key] ?? null;
+    }
+
+    /**
```
Comment (ihor-sviziev) [inline lib/internal/Magento/Framework/App/DeploymentConfig.php:259]:
```suggestion
    /**
     * Returns data by key
     *
```

**Review by ihor-sviziev — CHANGES_REQUESTED** [review]

**Comment by ihor-sviziev** [comment]
@magento run all tests

**Review by ihor-sviziev — APPROVED** [review]

**Comment by engcom-Alfa** [comment]
@magento run all tests

**Comment by engcom-Lima** [comment]
@magento run Functional Tests CE , Functional Tests EE

**Comment by engcom-Lima** [comment]
✔️ QA Passed

**Manual testing scenario:**

1. Installed a fresh Magento2.4-develop instance, and make sure that is working fine
2. Delete the database completely and re-create the same database name (That will be empty database as database deleted & re created)
3. Remove the file app/etc/env.php
4. Re-install the with the database creation command i.e. `php bin/magento setup:install --backend-frontname="admin" --session-save="files" --db-host="localhost" --db-name="****" --db-user="root" --db-password="" --base-url="****" --base-url-secure="https://local.magento24develop.com/" --admin-user="admin" --admin-password="admin123" --admin-email="admin@admin.com" --admin-firstname="Admin" --admin-lastname="Admin"`

Before: ✖️ Getting an error Connection default not defined while executing the installation command
Screenshot 2022-09-19 at 5 35 23 PM

After: ✔️ Successfully installing using the command
Screenshot 2022-09-19 at 5 39 07 PM

No additional manual testing is required.

**Comment by ishakhsuvarov** [comment]
@magento run all tests

**Comment by engcom-Lima** [comment]
@magento run Functional Tests EE, Integration Tests, WebAPI Tests

**Comment by engcom-Lima** [comment]
@magento run Functional Tests EE, Integration Tests

**Comment by engcom-Lima** [comment]
@magento run Integration Tests

**Comment by engcom-Echo** [comment]
@magento run all tests

**Comment by engcom-Echo** [comment]
@magento run Integration Tests, Functional Tests CE, Functional Tests EE

**Comment by engcom-Echo** [comment]
@magento run Functional Tests B2B

**Pull request merged**